### PR TITLE
gc: Fix GcKeys task doesn't work when called with multiple keys

### LINKED ISSUE #11217: TiKV GcKeys task doesn't work when called with multiple keys (at least in 5.1 but I think for everything)
## Bug Report

In gc worker, https://github.com/tikv/tikv/blob/master/src/server/gc_worker/gc_worker.rs#L340 the GcKeys GC task when called with multiple keys tries to get the list of regions overlapping the key range for the task and then do a sorted merge of the keys and the space encompassed by the regions. However, it fails to prepend the keys with the data prefix ('z') when passing them to get_regions_in_range, get_regions_in_range doesn't prepend the keys, and so the keys form an invalid range and no regions are ever returned. This makes garbage collection of delete markers not happen.

To verify the issue I added some logging statements in a branch: https://github.com/frew/tikv/pull/1/files#diff-fad0fef4b49a4159243096a9212032ae40e8b88cd0deb3c60df42a3dbdc639dfR385 - an example line from the first logging statement showing the issue in our cluster:
```
[2021/11/03 06:46:28.397 +00:00] [INFO] [region_info_accessor.rs:385] ["gc get_regions_in_range"] [start_key="[116, 128, 0, 0, 0, 0, 0, 0, 255, 159, 95, 114, 128, 0, 0, 12, 81, 255, 152, 74, 144, 0, 0, 0, 0, 0, 250]"] [first_region="Some(([122, 116, 128, 0, 0, 0, 0, 0, 0, 255, 19, 0, 0, 0, 0, 0, 0, 0, 248], 53))"] [regions_size=126845]
```

(note the prefix 122 on the first_region but not on the start_key)

I suggest adding data_key() to https://github.com/tikv/tikv/blob/master/src/server/gc_worker/gc_worker.rs#L338 and the following line, but defer to your expertise.

### What version of TiKV are you using?
5.1.2

### What operating system and CPU are you using?
Linux on GCP n2d

### PR DESCRIPTION
**Pull request by Connor1996**
### What problem does this PR solve?

Issue Number: close #11217  

Problem Summary:

The problem is because `gc_worker` passes the range with z-prefixed keys to get regions in the range, whereas `RegionInfoProvider` actually accepts keys without z-prefixed. So it causes `get_regions_in_range` returns incorrect results leading to unexpected behavior of handling `GcKeys` task.

Apart from that, find two more bugs:
- the implementation of `gc_keys` wrongly assume the start key and end key of region are z-prefixed
- `handle_get_regions_in_range` doesn't returned all covered regions of the keys.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

- Add a wrapper `RangeKey` to remove confusing z-prefixed keys in `RegionInfoProvider` 
- Fix the the implementation of `gc_keys`
- Fix `handle_get_regions_in_range` logic
- Add a metrics to show how many mvcc deletion version is handled
image

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test: using an update heavy workload, and do select to trigger scan, the key_skipped_count(due to too many mvcc deleted versions left) is greatly reduced.

Before
image

After
image

### Release note 

```release-note
Fix mvcc deletion may not be dropped by compaction filter GC 
```

### CODE REVIEW

**Comment by frew** [comment]
I think the RangeKey abstraction + not using data keys there is a useful change and looks good to me (although not the expert).

I don't believe this fixes the GcKeys problem because the logic in `gc_keys` expects
`get_regions_in_range` to return all regions *encompassing* the range (i.e. if you had three regions ['', 'b'), ['b', 'd'), ['d',) and `gc_keys` calls `get_regions_in_range('a', 'c')` for the function to work correctly `get_regions_in_range` needs to return both ['', 'b'), ['b', 'd') and similarly if the range is entirely enclosed in a single region). I'm *not* sure if CompactionGuard expects the same (and is also broken) or if it expects the current behavior.

**Review by yiwu-arbug — COMMENTED** [review]
Looking good.

Can we add an integration test making sure compaction filter cleanup MVCC deletion marker?

File: components/raftstore/src/coprocessor/region_info_accessor.rs [hunk 1]
```diff
@@ -67,7 +67,80 @@ impl RegionInfo {
...
 
 type RegionsMap = HashMap<u64, RegionInfo>;
-type RegionRangesMap = BTreeMap<Vec<u8>, u64>;
+type RegionRangesMap = BTreeMap<RangeKey, u64>;
+
+// RangeKey is a wrapper used to unify the comparsion between region start key
+// and region end key. Region end key is special as empty stands for the infinite,
+// so we need to take special care for cases where the end key is empty.
+#[derive(Clone, Debug)]
+pub struct RangeKey {
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply components/raftstore/src/coprocessor/region_info_accessor.rs:76]: @frew cool! Didn't notice that enum variants are ordered by their top-to-bottom discriminant order

File: components/raftstore/src/coprocessor/region_info_accessor.rs [hunk 2]
```diff
@@ -67,7 +67,80 @@ impl RegionInfo {
...
+    key: Vec<u8>,
+    is_end_key: bool,
+}
+
+impl RangeKey {
+    pub fn from_start_key(key: Vec<u8>) -> Self {
+        Self {
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply components/raftstore/src/coprocessor/region_info_accessor.rs:83]: start_key may be empty which stands for the -infinite

[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply components/raftstore/src/coprocessor/region_info_accessor.rs:83]: ok

**Review by yiwu-arbug — APPROVED** [review]
LGTM

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**